### PR TITLE
Allow the player to track newly learned/unviewed recipes

### LINKED ISSUE #49829: hidden unexplained crafting menu
### Describe the bug

There is a menu in crafting I got to. I don't know what it's for and I don't know how was I supposed to know how to get there.

### Steps To Reproduce

To get to the menu:
1. Probably regular character that played a bit and has some materials for fabrication recipes.
2. Open crafting menu.
3. Search "p:fab" (search for recipes using fabrication as primary skill).
4. Press the left arrow twice.

### Expected behavior

Explanation on what the menu is once I get there. How are the crafting recipes sorted?

Explanation on how to get to that menu e.g. "folders"/"cards" that are at the top that switch by pressing tab or arrows like everywhere else.

### Screenshots

Screenshot_2

### Versions and configuration

- OS: Windows 10
- Game Version: 19c4b2f
- Graphics version: Tiles
- Ingame language: English
- Mods loaded: dda, no_npc_food, package_bionic_proffessions

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Qrox**
#### Summary
Features "Allow the player to track newly learned/unviewed recipes"

#### Purpose of change
Adds an "unread" status to the recipes which can be used to track new recipes that the player has learned but not viewed yet.

#### Describe the solution
- [x] Add a new `read_recipes` member to `uistate`
- [x] Implement display/toggling/sorting of unread recipes in the crafting gui
- [x] Add an option to disable the feature
- [x] Playtesting

#### Testing

recipe

- [x] The `NEW!` marker is displayed for unread recipes
- [x] Tabs with unread recipes are marked with superscript plus signs
- [x] The unread status is cleared by selecting recipe with UP/DOWN, selecting batch, or crafting
- [x] The unread status is toggled by pressing v
- [x] Listed/filtered/all recipes can be marked as read by pressing shift+v
- [x] Unread recipes can be displayed at the top of the list by pressing u
- [x] The feature can be turned off in the options menu
- [x] Fixes #49829

#### Additional context
I chose the keybindings arbitrarily. Maybe someone can suggest better keybindings?

Also some ideas for further PRs:
- Maybe we can prompt the player when they learn new recipes from skill-up or reading books.
- A similar system can be made for the construction menu.

### CODE REVIEW

**Comment by pjf** [comment]
I see that only 'playtesting' is left, so I'm adding this to my local build now.

I've wanted something like this for *so long*. Thank you for making my dreams come true!

**Comment by pjf** [comment]
Reporting in that after a few hours of play it's all worked as expected! :tada:

The only thing I've found odd is that the same hotkey ('v') is showing for mark read (which it operates as), and mark all read (which it does not). The local keybindings display both functions being identically bound.

**Comment by Qrox** [comment]
Thanks for reporting! Fixed it by assigning captial letter `V` to the mark all action.

**Comment by pjf** [comment]
On the playtesting front I haven't tested the latest commit addressing #49829, but everything up to and including b19ccca has been working *flawlessly*. Thank you again!

**Comment by leemuar** [comment]
Awesome addition, thanks!

**Comment by Qrox** [comment]
Played with the feature for a while and it works fine, should be ready to merge.

**Comment by Brambor** [comment]
Amazing feature, what about the 2 failing tests? Could it be fixed & merged ASAP? Can I help?

**Comment by Qrox** [comment]
The unit test failure is likely unrelated. The clang-tidy error is about a function being too complex, but to be honest that function was very complex to begin with. I'll try to fix it.

**Comment by Brambor** [comment]
Out of curiosity, did https://github.com/CleverRaven/Cataclysm-DDA/commit/a17ccb1888b5dde41c1cf8936f51d2bf69c23916 fix tthis issue? Would merging master fix the tests?

**Comment by Qrox** [comment]
That commit doesn't look related to the errors. And I don't think the clang-tidy error is fixed in master -- that error was introduced because I modified the function in this PR.

**Comment by kevingranade** [comment]
I did a trivial fix to get the function under the limit by extracting context generation to a helper function.

**Comment by Brambor** [comment]
Now the tests fail on astyle.

When I checkout last commit, make astyle and look for changes, I see no changes.
```
git checkout 38a7d7a408b7e7d5fc4d71f8f027ca8e793d3441 #kevingrades commit
make astyle
git status
```

When I checkout to some recent "fix astyle" commit somewhere else, then `checkout HEAD~1`, make & status, I see the expected changes.

TL;DR: I wanted to help, now I am even more confused on what's failing.

**Comment by kevingranade** [comment]
It's failing because a badly styled commit made it into the main branch.
